### PR TITLE
Prevent API groups from growing stale

### PR DESCRIPTION
**Pull request by fabianvf**
@willthames @asetty if you'd like to take a look

### CODE REVIEW

**Comment by willthames** [comment]
What is the difference between the existing refresh mechanism and the one resulting from this change? It's not obvious to me how the refresh is triggered.

File: openshift/dynamic/client.py [hunk 1]
```diff
@@ -657,13 +657,16 @@ def parse_api_groups(self, request_resources=False):
...
+                    resource_group = self._cache.get('resources', {}).get(DISCOVERY_PREFIX, {}).get(group['name'], {}).get(version)
                     preferred = version_raw == group['preferredVersion']
-                    resources = {}
+                    resources = resource_group.resources if resource_group else {}
                     if request_resources:
                         resources = self.get_resources_for_api_version(DISCOVERY_PREFIX, group['name'], version, preferred)
                     new_group[version] = ResourceGroup(preferred, resources=resources)
                 groups[DISCOVERY_PREFIX][group['name']] = new_group
-            self._cache['resources'] = groups
+                self._cache['resources'].update(groups)
```
Comment (asetty) [inline openshift/dynamic/client.py]: Wouldn't doing the dictionary update every iteration of the loop be slower than just setting self._cache['resources'] at the end?

**Comment by fabianvf** [comment]
This makes the api_groups refresh when accessed, so that it's always the most up to date list of resources available in the cluster. Before it would only refresh when you attempted to access a resource that did not exist in the cache. Since updating api_groups is just a single API call (with the lazy discoverer at least), I thought this behavior would be more useful than requiring you to manually refresh the cache or miss a resource to get the new groups to show up. We hit this in one of our operators, where it created a CRD and later verified that the API group existed- because api_groups didn't refresh, the check failed.

File: openshift/dynamic/client.py (hunk 1 above)
Comment (fabianvf) [reply openshift/dynamic/client.py]: woops, typo, that was supposed to be one level out

**Comment by djzager** [comment]
I think I tested this correctly:

```
#!/usr/bin/env python
import yaml
import pprint
from kubernetes import client, config
from openshift.dynamic import DynamicClient

k8s_client = config.new_client_from_config()
dyn_client = DynamicClient(k8s_client)
api_groups = dyn_client.resources.api_groups
for res in api_groups:
    print(res)
```

```
➜  openshift-restclient-python git:(pr/264) ✗ ./example.py

certificates.k8s.io
events.k8s.io
authorization.k8s.io
apiextensions.k8s.io
autoscaling
apps
batch
apiregistration.k8s.io
networking.k8s.io
extensions
storage.k8s.io
policy
coordination.k8s.io
rbac.authorization.k8s.io
scheduling.k8s.io
admissionregistration.k8s.io
authentication.k8s.io
```

Then I create my CRD

```
➜  openshift-restclient-python git:(pr/264) ✗ ./example.py

examples.djzager.io
certificates.k8s.io
events.k8s.io
authorization.k8s.io
apiextensions.k8s.io
autoscaling
apps
batch
apiregistration.k8s.io
networking.k8s.io
extensions
storage.k8s.io
policy
coordination.k8s.io
rbac.authorization.k8s.io
scheduling.k8s.io
admissionregistration.k8s.io
authentication.k8s.io
```

**Comment by djzager** [comment]
And I can show it failing on `master`:

```
(osrcp) ➜  openshift-restclient-python git:(master) ✗ ./example.py
Traceback (most recent call last):
  File "./example.py", line 10, in <module>
    for res in api_groups:
TypeError: 'instancemethod' object is not iterable
```

**Comment by fabianvf** [comment]
@willthames did that clear up your reservations?

**Pull request merged**